File: determine_strings_close.cpp
#include "leetcode.h"

class Solution {
public:
  bool closeStrings(string word1, string word2) {
    if (word1.size() != word2.size())
      return false;
    vector<int> c1(26), c2(26);
    for (auto i = 0; i < word1.size(); ++i) {
      ++c1[word1[i] - 'a'];
      ++c2[word2[i] - 'a'];
    }
    if (!equal(begin(c1), end(c1), begin(c2), end(c2),
               [](int a, int b) { return (bool)a == bool(b); }))
      return false;
    sort(begin(c1), end(c1));
    sort(begin(c2), end(c2));
    return c1 == c2;
  }
};

int main() {
  Solution obj;
  if (obj.closeStrings("abc", "bca"))
    cout << "true";
  else
    cout << "false";
}

File: other/determine_strings_close.rs
struct Solution;

impl Solution {
    pub fn close_strings(word1: String, word2: String) -> bool {
        let (mut c1, mut c2) = ([0; 26], [0; 26]);
        for &b in word1.as_bytes() {
            c1[(b - b'a') as usize] += 1;
        }
        for &b in word2.as_bytes() {
            c2[(b - b'a') as usize] += 1;
        }
        if (0..26).any(|i| (c1[i] > 0) != (c2[i] > 0)) {
            return false;
        }
        c1.sort_unstable();
        c2.sort_unstable();
        c1 == c2
    }
}

fn main() {
    let word1 = String::from("abc");
    let word2 = String::from("bca");
    print!("{}", Solution::close_strings(word1, word2));
}
